### ISSUE TITLE
Should emit `used-before-assignment` for unused type annotation after the same name appears multiple times in a comprehension

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#5654]
### Bug description

```python
"""https://github.com/PyCQA/pylint/issues/5654"""
from typing import Any
def func():
    """Should raise used-before-assignment on the last line"""
    my_none: Any
    _ = [print(sep=my_none, end=my_none) for my_none in (None, None)]
    print(my_none)
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
On `main`, it's raised on the wrong line (the comprehension line):
test.py:5:37: E0602: Undefined variable 'my_none' (undefined-variable)

--------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 10.00/10, -10.00)

After #5651, no warnings at all:

$ pylint ../test.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
```

### Expected behavior

test.py:7:37: E0601: Using variable 'my_none' before assignment (used-before-assignment)

### Pylint version

```shell
pylint 2.13.0-dev0
astroid 2.9.2
Python 3.10.1 (v3.10.1:2cd268a3a9, Dec  6 2021, 14:28:59) [Clang 13.0.0 (clang-1300.0.29.3)]
```

**Comment by jacobtylerwalls** [comment]
Edited to suggest `used-before-assignment` because the Python interpreter produces:
```
UnboundLocalError: local variable 'my_none' referenced before assignment
```
